src/rust/catnap/win/overlapped.rs

```diff
@@ -64,37 +56,31 @@ pub struct OverlappedResult {
 
 /// Data required by the I/O completion port processor to process I/O completions.
 #[repr(C)]
-struct OverlappedCompletion {
+struct OverlappedCompletion<S: Unpin> {
```

Comment:
The semantics of Unpin are still a little unclear to me. From the Rust docs:
> Implementing the Unpin trait for T expresses the fact that T is pinning-agnostic: it shall not expose nor rely on any pinning guarantees.

The type `SocketOpState` itself doesn't care about pinning, e.g. it's not self-referential; however, it can't be used for its intended purpose without pinning, because e.g. the address memory pointer on AcceptEx must be pinned during the call. Implementing `Unpin` in this case seems to just make the pinning use more implicit. This has the benefit of making the code easier to write, with fewer unsafe blocks, but the written code is less expressive and therefore more error-prone to future development. Not sure this merits a rewrite of what's here, just trying to build my own understanding.

Reply:
I guess in my first attempt, the lack of Unpin was explicit, because every time I referenced the State, I wanted to stop and think whether it was valid, then write a justification for why such an access was valid at that point. But maybe that's an overuse of unsafe?

Reply:
Not going to block on this.